### PR TITLE
Wandb callback

### LINKED ISSUE #46647: Weights & Biases Callback for supporting logging to Weights & Biases
### 需求描述 Feature Description

In this issue, I am proposing creating a new callback for logging metrics to Weights & Biases. This will be added [here](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/hapi/callbacks.py) with the rest of the callbacks. I would love to implement this as well!

The added functionality would include logging metrics to a W&B dashboard along with saving model checkpoints as W&B artifacts with only 2 extra lines of code for the users. 

This will give users the ability to track their model training with plots like this

Screenshot 2022-09-29 at 2 30 13 PM

This will be accompanied by being able to track and version saved checkpoints

Screenshot 2022-09-29 at 2 30 59 PM

W&B also automatically tracks all system metrics during training along with model config for reproducibility.

Screenshot 2022-09-29 at 2 32 50 PM

Screenshot 2022-09-29 at 2 32 28 PM

A W&B integration already exists in other libraries supported by PaddlePaddle like [PaddleOCR](https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.6/ppocr/utils/loggers/wandb_logger.py) and [PaddleDetection](https://github.com/PaddlePaddle/PaddleDetection/blob/release/2.5/ppdet/engine/callbacks.py#L291) along with other popular libraries like MMEngine, MMDetection and more.

Adding this callback will require adding any addition requirements to the paddlepaddle python package either! I would love to hear the mantainers' thoughts on this.

### 替代实现 Alternatives

_No response_

### PR DESCRIPTION
**Pull request by manangoel99**
### PR types
New features

### PR changes
Others

### Describe
This PR fixes #46647 by adding a new `WandbCallback` which automatically logs training, validation and systems metrics to a wandb dashboard like the one shown here.   

Screenshot 2022-10-11 at 10 42 47 PM
Screenshot 2022-10-11 at 10 43 06 PM

It also works out of the box for distributed training. Script to reproduce these plots is given in the comments.

Another proposed change is adding the option to add a custom error message to the `paddle.utils.try_import` function.

### CODE REVIEW

**Comment by manangoel99** [comment]
```
import paddle
import paddle.vision.transforms as T
from paddle.static import InputSpec

inputs = [InputSpec([-1, 1, 28, 28], 'float32', 'image')]
labels = [InputSpec([None, 1], 'int64', 'label')]

transform = T.Compose([
    T.Transpose(),
    T.Normalize([127.5], [127.5])
])

train_dataset = paddle.vision.datasets.MNIST(mode='train', transform=transform)
eval_dataset = paddle.vision.datasets.MNIST(mode='test', transform=transform)

net = paddle.vision.models.LeNet()
model = paddle.Model(net, inputs, labels)

optim = paddle.optimizer.Adam(0.001, parameters=net.parameters())
model.prepare(optimizer=optim,
            loss=paddle.nn.CrossEntropyLoss(),
            metrics=paddle.metric.Accuracy())

callback = paddle.callbacks.WandbCallback(project='paddle_mnist')
model_callback = paddle.callbacks.ModelCheckpoint('./checkpoint')
model.fit(train_dataset, eval_dataset, batch_size=64, callbacks=[callback, model_callback], epochs=5)

```

**Comment by manangoel99** [comment]
@heavengate @jzhang533 It would be awesome if I could get a review on this PR

**Comment by AndPuQing** [comment]
You may need to add a unit test, like https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/tests/test_callback_visualdl.py

File: python/paddle/hapi/callbacks.py [hunk 1]
```diff
@@ -957,6 +957,175 @@ def on_eval_end(self, logs=None):
...
+
+        pip install wandb
+        wandb login
+
+    Args:
+        project(str, optional): Name of the project. Default: uncategorized
+        entity(str, optional): Name of the team/user creating the run. Default: Logged in user
+        name(str, optional): Name of the run. Default: randomly generated by wandb
+        dir(str, optional): Directory in which all the metadata is stored. Default: `wandb`
+        offline(str, optional): Can be "online", "offline" or "disabled". Default: "online".
```
Comment (jzhang533) [inline python/paddle/hapi/callbacks.py]: this should be `mode` arg?

**Review by jzhang533 — COMMENTED** [review]
is it possible to have a testcase to cover this feature?

File: python/paddle/hapi/callbacks.py (hunk 1 above)
Comment (manangoel99) [reply python/paddle/hapi/callbacks.py]: Oh yes! Fixing it now!

**Comment by manangoel99** [comment]
For implementing a unit test, the wandb python library would have to be added to the requirements.txt file for the package. Is that an acceptable change?

**Comment by AndPuQing** [comment]
> For implementing a unit test, the wandb python library would have to be added to the requirements.txt file for the package. Is that an acceptable change?

add it in https://github.com/PaddlePaddle/Paddle/blob/develop/python/unittest_py/requirements.txt . just import it when utest.

File: python/paddle/hapi/callbacks.py [hunk 2]
```diff
@@ -957,6 +957,175 @@ def on_eval_end(self, logs=None):
...
+                               name=name,
+                               entity=entity,
+                               dir=dir,
+                               mode=mode,
+                               job_type=job_type)
+
+        self._run = None
+        self.wandb_args.update(**kwargs)
+
+        _ = self.run
```
Comment (heavengate) [inline python/paddle/hapi/callbacks.py:1074]: What does this line do? `_` is usually used as unused variable.

File: python/paddle/tests/test_callback_wandb.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline python/paddle/tests/test_callback_wandb.py]: 2020 -> 2022

File: python/paddle/tests/test_callback_wandb.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+from paddle.fluid.framework import _test_eager_guard
+
+
+class MnistDataset(MNIST):
+
+    def __len__(self):
+        return 512
+
+
+class TestCallbacks(unittest.TestCase):
```
Comment (heavengate) [inline python/paddle/tests/test_callback_wandb.py]: TestCallbacks -> TestWandbCallbacks

File: python/paddle/hapi/callbacks.py (hunk 2 above)
Comment (manangoel99) [reply python/paddle/hapi/callbacks.py:1074]: This line initializes the wandb run if a run is not already in progess in which case it uses the ongoing run by calling this method

https://github.com/manangoel99/Paddle/blob/8353b6996d4bc157a6dd416ecbadb35df3c75346/python/paddle/hapi/callbacks.py#L1039-L1053

**Comment by heavengate** [comment]
There are some errors in CI tests， you can click Details for logs:
截屏2022-10-12 下午5 08 24
1. Static Check: CI with extract and run examples code from Python comments,  it seems paddle is not imported in exmaple codes
截屏2022-10-12 下午5 11 14
2.  test_callback_wandb run failed or timeout in Mac/Windows, please check locally
截屏2022-10-12 下午5 13 12

**Comment by manangoel99** [comment]
@heavengate Thank you for the detailed review! I'll look into these errors and update the PR

**Comment by manangoel99** [comment]
@heavengate for the Mac build I'm getting an import error even though the `wandb` package has been added to the `python/unittest_py/requirements.txt` file. I'm not sure how to fix this.

Screenshot 2022-10-18 at 4 06 29 PM

Also for the Static check for the example codes, it is giving an ImportError that `import paddle` has not been called even though the line is [here](https://github.com/manangoel99/Paddle/blob/6311dfcca85461dc0a6495374d6722b0dcc8a3b4/python/paddle/hapi/callbacks.py#L983)

Could you guide me on how to fix these

**Comment by heavengate** [comment]
> @heavengate for the Mac build I'm getting an import error even though the `wandb` package has been added to the `python/unittest_py/requirements.txt` file. I'm not sure how to fix this.
> 
> Screenshot 2022-10-18 at 4 06 29 PM
> 
> Also for the Static check for the example codes, it is giving an ImportError that `import paddle` has not been called even though the line is [here](https://github.com/manangoel99/Paddle/blob/6311dfcca85461dc0a6495374d6722b0dcc8a3b4/python/paddle/hapi/callbacks.py#L983)
> 
> Could you guide me on how to fix these

It seems `wandb` install failed due to environment, I rerun the CI and it is already passed

**Comment by luotao1** [comment]
`test_callback_wandb (Timeout)` in PR-CI-Coverage
```
2022-10-20 15:55:18     Test #464: test_callback_wandb ..............***Timeout  15.00 sec
```
How much time of test_callback_wandb? If you need much more time, you could `set_tests_properties` like：  
https://github.com/PaddlePaddle/Paddle/blob/ef80073cb82cf4341ff9ebb547ee2694f49186cd/python/paddle/tests/CMakeLists.txt#L59

**Comment by manangoel99** [comment]
@luotao1 Thank you for the suggestion! I've added that

File: python/paddle/tests/CMakeLists.txt [hunk 5]
```diff
@@ -66,6 +66,7 @@ set_tests_properties(test_vision_models PROPERTIES TIMEOUT 120)
 set_tests_properties(test_dataset_uci_housing PROPERTIES TIMEOUT 120)
 set_tests_properties(test_dataset_imdb PROPERTIES TIMEOUT 300)
 set_tests_properties(test_pretrained_model PROPERTIES TIMEOUT 600)
+set_tests_properties(test_callback_wandb PROPERTIES TIMEOUT 300)
```
Comment (luotao1) [inline python/paddle/tests/CMakeLists.txt]: Since the elapsed time of `test_callback_wandb` are less than 30 seconds, thus, `TIMEOUT 300`->`TIMEOUT 60`?
```
2022-10-20 18:46:06     Test #463: test_callback_wandb ..............   Passed   28.08 sec
2022-10-20 18:46:06     Start 463: test_callback_wandb
2022-10-20 18:46:29     Test #463: test_callback_wandb ..............   Passed   23.44 sec
2022-10-20 18:46:29     Start 463: test_callback_wandb
2022-10-20 18:46:53 1/1 Test #463: test_callback_wandb ..............   Passed   23.59 sec
```

File: python/paddle/utils/lazy_import.py [hunk 6]
```diff
@@ -32,9 +32,10 @@ def try_import(module_name):
...
-            "Failed importing {}. This likely means that some paddle modules "
-            "require additional dependencies that have to be "
-            "manually installed (usually with `pip install {}`). ").format(
-                module_name, install_name)
+        if err_msg is None:
+            err_msg = (
+                "Failed importing {}. This likely means that some paddle modules "
+                "require additional dependencies that have to be "
+                "manually installed (usually with `pip install {}`). ").format(
+                    module_name, install_name)
```
Comment (luotao1) [inline python/paddle/utils/lazy_import.py]: image

image
The change here doesn't pass the python coverage test, could you add a unit-test or revert the change (use the default error message)?

File: python/unittest_py/requirements.txt [hunk 7]
```diff
@@ -17,3 +17,4 @@ numpy>=1.20,<1.22; python_version >= "3.7"
 autograd==1.4
 librosa==0.8.1
 parameterized
+wandb
```
Comment (luotao1) [inline python/unittest_py/requirements.txt]: How about specific the version of wandb?

File: python/paddle/utils/lazy_import.py (hunk 6 above)
Comment (manangoel99) [reply python/paddle/utils/lazy_import.py]: I added aa unit test. Thank you for the suggestion

File: python/paddle/tests/CMakeLists.txt (hunk 5 above)
Comment (manangoel99) [reply python/paddle/tests/CMakeLists.txt]: I've made the requested change

**Comment by luotao1** [comment]
LGTM, please solve the code conflicts

**Review by heavengate — DISMISSED** [review]

**Comment by manangoel99** [comment]
@luotao1 I've fixed the merge conflicts and updated the PR

**Review by luotao1 — APPROVED** [review]
LGTM

**Review by jzhang533 — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM

**Review by jeff41404 — APPROVED** [review]
LGTM

**Comment by luotao1** [comment]
@manangoel99 Thanks very much for your contribution!
- We need a Chinese API document for `paddle.callbacks.WandbCallback`, like [`paddle.callbacks`](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/callbacks/Overview_cn.html). The source code of the Chinese API document should be put in  [docs#api/paddle/callbacks](https://github.com/PaddlePaddle/docs/tree/develop/docs/api/paddle/callbacks). Since we find https://docs.wandb.ai/v/zh-hans/ have a Chinese version, could you introduce a developer to contribute the Chinese API document for `paddle.callbacks.WandbCallback`?
- We have a bi-week online Paddle framework contributor meetup, if you have interest to give a talk to introduce your project at this event, please let us know.

**Comment by manangoel99** [comment]
@luotao1 Sure! We can add the Chinese API documentation as well. Can you also tell me where the english documentation would go so that I can write up a more detailed bit of documentation or is it built from the docstring?

I would love to be a part of the contributor meet up and talk to the people about what this PR does and how Weights & Biases is useful

**Comment by luotao1** [comment]
> Can you also tell me where the English documentation would go so that I can write up a more detailed bit of documentation or is it built from the docstring?

Yes, English documentation is built from the docstring, like [`paddle.callbacks.Callback`](https://www.paddlepaddle.org.cn/documentation/docs/en/develop/api/paddle/callbacks/Callback_en.html) in our website. Our website needs some time to build the latest develop branch, we will tell you if `paddle.callbacks.WandbCallback` is shown on the website.

> I would love to be a part of the contributor meet up and talk to the people about what this PR does and how Weights & Biases is useful

Got it. We will discuss the details of the meetup event internally first and then discuss it with you.

**Comment by luotao1** [comment]
@manangoel99 Our website updates, you can see https://www.paddlepaddle.org.cn/documentation/docs/en/develop/api/paddle/callbacks/WandbCallback_en.html

**Comment by luotao1** [comment]
>  We will discuss the details of the meetup event internally first

How about: 7-8 p.m., November 24, Beijing time ?
Tencent Meeting: https://apps.apple.com/cn/app/tencent-meeting/id1484048379?l=en

**Comment by manangoel99** [comment]
@luotao1 sounds good! Thank you!
